UI: Add ability to reorder filters by drag & drop

cg2121 opened the pull request:
### Description
This adds the ability for filters to be reordered by drag & drop.

### Motivation and Context
The scenes and sources can be reordered this way, so it makes sense for filters as well.
https://ideas.obsproject.com/posts/369/re-order-filter-effects

### How Has This Been Tested?
I tested this by adding filters and dragging and dropping them. Everything worked as expected.

### Types of changes
- New feature (non-breaking change which adds functionality)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.

cg2121 committed:
UI: Add ability to reorder filters by drag & drop

This adds the ability for filters to be dragged and dropped
to be reordered, similar to scenes and sources.

RytoEX commented on UI/forms/OBSBasicFilters.ui:
```diff
@@ -414,7 +432,7 @@
           <widget class="QDialogButtonBox" name="buttonBox">
            <property name="standardButtons">
-            <set>QDialogButtonBox::Reset|QDialogButtonBox::Close</set>
+            <set>QDialogButtonBox::Close|QDialogButtonBox::Reset</set>
```
Did these buttons get reordered for some reason?  Does this affect UI or tab order?

cg2121 replied:
QT designer reordered them. They still show up normally.

WizardCM commented:
Just tested this myself, it seemed to have worked about half the time. Even less likely to work if you have only just added the filters. I tested with two filters, and had trouble moving the top filter to the bottom, and vice versa - for both A/V Filters & Effect Filters.

cg2121 commented:
@WizardCM I've now made some changes to the code. Could you check if it works for you now?

cg2121 commented:
The filters backend needs to be rewritten for this to work properly. It works just fine if you are only using async or effect filters. If you use both types, that is when things start not to work. Both async and effect filters are on the same filters array, so finding the correct index to move a filter to is impossible. This can be fixed if the async and effect filters are put in separate arrays.

I will close for the time being, until this issue is fixed.

cg2121 commented:
Reopening, as I have found a way to make it work.

WizardCM commented:
Errors out in Debug mode when starting drag.

![image](https://user-images.githubusercontent.com/941350/115100740-f120fa80-9f81-11eb-8a02-4c556a33439a.png)

I can confirm in RelWithDebInfo that this seems to work perfectly, for both video sources & audio sources.

cg2121 commented:
This needs a custom list model to work properly, similar to the source tree, so I am closing for now.

cg2121 commented:
Reopened, as I figured out a way to simply do this.

Turns out it is a Qt bug that causes the items to disappear (https://bugreports.qt.io/browse/QTBUG-100128). This was fixed by using the following workaround https://stackoverflow.com/questions/74263946/widget-inside-qlistwidgetitem-disappears-after-internal-move/75832468#75832468

And with the introduction of UUIDs, the QDataStream now works properly with filters, as before the OBSSource QDataStream would work with source names, so filters couldn't be accessed with it as they were private sources.

@WizardCM could you confirm this works for you now also?

WizardCM commented:
Verified I couldn't reproduce any of the previously reported bugs. Everything seems to work correctly. Will review code separately.

cg2121 committed:
UI: Add QDataStream opreators for OBSSource

Allows serializing/deserializing sources in and out of QDataStreams via
their UUID

cg2121 committed:
libobs: Add functions to get/set the index of filters
